server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java

```diff
@@ -119,27 +109,66 @@ public TemporalAccessor parse(String input) {
         if (Strings.isNullOrEmpty(input)) {
             throw new IllegalArgumentException("cannot parse empty date");
         }
-        return parser.parse(input);
+
+        try {
+            return doParse(input);
+        } catch (DateTimeParseException e) {
+            throw new IllegalArgumentException("failed to parse date field [" + input + "] with format [" + format + "]", e);
```

Comment:
Actually the javadoc for the interface would should state RuntimeException to be absolutely correct, but it would be very imprecise.
`JodaDateFormatter` can throw `java.time.DateTimeException` as well as `UnsupportedOperationException` and `IllegalArgumentException`
`JavaDateFormatter` only `IllegalArgumentException`. `IllegalArgumentException` is stated in the underlying doc for `Format.parseObject` but I don't expect this to be thrown as we already check for input not being empty and provide the position of parsing.

I will refactor the javadoc to state that `IllegalArgumentException` is expected to be thrown when parsing fails (that applies in both cases).  But I wonder if we should ommit the `DateTimeException` and `UnsupportedOperationException` (thrown only by joda)

Reply:
I think there is no point in digging through the complete implementation of Java time and Joda time in order to come up with a Javadoc that documents all possible exceptions but rather document the ones that a user can "reasonably" expect when a date cannot be parsed which is in our case `java.time.DateTimeException` from `JodaDateFormatter` and `IllegalArgumentException` (from both) so they know how to deal with them. 

IMHO it is ok to document both of them on the interface. Although one of them is in fact thrown by only one of the implementations, the exception type itself is available in the JDK and thus generic enough and we should expect users to program against the interface.